ctcusc/RiverLA
# Wiencko/fix tagging

## Ticket Information


When the PersonCreated endpoint is called, the interest category tags that label the volunteers aren't immediately added to the person. Since the tags are used to match volunteers with organizations, and there are no tags associated with the PersonCreated endpoint, the email is coming in with no organization listed :( 

My ticket was to fix this issue.

**Link to Notion**

https://www.notion.so/ctc/RiverLA-52449746357b4216b1c072e7995ff72d?p=91b0a3d255a84650b03e11587e58b4bd

**Short Description**

To fix the problem, instead of directly acting on the object that we receive when PersonCreated is called (who has no tags since that happens after they are created), we wait for 60 seconds before using Nationbuilder's API to get current information on that person using their ID that we received during the PersonCreated call. We can then use the updated information on tags to continue with the organization selection and email sending code that has already been written.

## Manual Test Plan

I manually by running dev (changing my env variable from development to production or anything other than development since it prevents emails from being sent) and adding a webhook to Nationbuilder under the admin/webhooks page in RiverLA's Nationbuilder [https://larivercorp.nationbuilder.com/admin/webhooks/new](https://larivercorp.nationbuilder.com/admin/webhooks/new), using the webhook URL that our code prints out after running dev. Then, I used a temp email (or my personal email) to create a volunteer on the RiverLA volunteer page. A minute later, I received an email with organizations corresponding to my organization type preferences. I did this several times in order to ensure the code was working. One thing to note: if our code refreshes, a new webhook is created and that needs to be added to Nationbuilder since the old one will stop working. Additionally, there is some time delay between the webhook being added to Nationbuilder and the webhook being fully operational (possibly if we queue up too many actions like adding many webhooks for rapid testing in development), so I usually give Nationbuilder five minutes after adding the webhook before I created new people, so I didn't get any false negatives on my testing (no email even when everything seems to be set up).

## PR Checklist

# Note

The tests have not been fully written because we've had some issues with test cases that have not been fully wrinkled out.

- [ ] Wrote ava tests
- [ ] Reviewed by **the tech lead and at least 2 other** @ctcusc/RiverLA team members
